doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,331 @@
+
+- LAG/LEAD (i.e., previous/next element of the window partition)
+    - We aim for only offset 1 in the first version, which is the default. Bigger offsets not seen in user queries yet, but shouldn’t be a problem to add support later.
+    - IGNORE NULLS should be supported. (already seen in a user query) (easy)
+- Window aggregations
+    - Small frames: small output changes for small input changes.
+    - Large frames: These are often impossible to support efficiently in a streaming setting, because small input changes might lead to big output changes. However, there are some aggregations which don’t necessarily result in big output changes even when applied with a large frame (Prefix Sum will automagically handle these cases efficiently):
+        - MIN/MAX, if usually the changed input element is not the smallest/largest.
+        - SUMming an expression that is often 0.
+        - Window aggregations with an UNBOUNDED PRECEDING frame are fine if changes happen mostly at the end of partitions
+            - e.g., OVER an `ORDER BY time` if new elements are arriving typically with fresh timestamps. Such OVER clauses are popular case in our [use case list](https://www.notion.so/Window-Functions-Use-Cases-6ad1846a7da942dc8fa28997d9c220dd).
+- FIRST_VALUE / LAST_VALUE / NTH_VALUE with various frames. These are similar to window aggregations.
+- ROW_NUMBER, RANK, DENSE_RANK, PERCENT_RANK, CUME_DIST, NTILE: These are impossible to implement efficiently in a streaming setting in many cases, because a small input change will lead to a big output change if the changed record is not near the end of the window partition. However, I can imagine some scenarios where the user knows some special property of the input data that ensures that small input changes will lead to small output changes, so she will use one of these functions and expect it to be efficient:
+    - If changes mostly come near the end of the window partition. For example, if there is an ORDER BY time, and new records usually have recent timestamps. (Prefix Sum will handle this fine.)
+    - If most changes are not record appearances or disappearances, but existing records changing in a way that they move only a little bit in the ordering. In this case, the output changes only for as many records, that got “jumped over” by the changing record. (Prefix Sum will handle this fine.)
+    - TopK is an important special case: This is when there is a `WHERE ROW_NUMBER() < k`
+        - Instead of doing Prefix Sum, we will transform this into our [efficient TopK implementation](https://www.notion.so/e62fe2b3d8354052ac7d0fe92be1e711).
+
+## Non-goals / Limitations
```

Comment:
I personally would not mix non-goals and limitations. Non-goals are things that you explicitly don't want to do, but limitations are part of the discussion of your solution.

Reply:
Hmm, yes. I separated them into two different sections.